hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheProxy.java

```diff
@@ -216,9 +219,63 @@ public V getAndPut(K key, V value, ExpiryPolicy expiryPolicy) {
     public void putAll(Map<? extends K, ? extends V> map, ExpiryPolicy expiryPolicy) {
         ensureOpen();
         validateNotNull(map);
-        // TODO implement batch putAll
+
+        int partitionCount = partitionService.getPartitionCount();
+
+        try {
+            List<Future> futures = new ArrayList<Future>(partitionCount);
+
+            // First we fill entry set per partition
+            List<Map.Entry<Data, Data>>[] entriesPerPartition = groupDataToPartitions(map, partitionCount);
+
+            // Then we invoke the operations and sync on completion of these operations
+            putToAllPartitionsAndWaitForCompletion(entriesPerPartition, futures, expiryPolicy);
+        } catch (Exception e) {
+            throw ExceptionUtil.rethrow(e);
+        }
+    }
+
+    private List<Map.Entry<Data, Data>>[] groupDataToPartitions(Map<? extends K, ? extends V> map,
+                                                                int partitionCount) {
+        List<Map.Entry<Data, Data>>[] entriesPerPartition = new List[partitionCount];
+
         for (Map.Entry<? extends K, ? extends V> entry : map.entrySet()) {
-            put(entry.getKey(), entry.getValue(), expiryPolicy);
+            K key = entry.getKey();
+            V value = entry.getValue();
+            validateNotNull(key, value);
+
+            Data keyData = serializationService.toData(key);
+            Data valueData = serializationService.toData(value);
+
+            int partitionId = partitionService.getPartitionId(keyData);
+            List<Map.Entry<Data, Data>> entries = entriesPerPartition[partitionId];
+            if (entries == null) {
+                entries = new ArrayList<Map.Entry<Data, Data>>();
+                entriesPerPartition[partitionId] = entries;
+            }
+
+            entries.add(new AbstractMap.SimpleImmutableEntry<Data, Data>(keyData, valueData));
+        }
+
+        return entriesPerPartition;
+    }
+
+    private void putToAllPartitionsAndWaitForCompletion(List<Map.Entry<Data, Data>>[] entriesPerPartition,
+                                                        List<Future> futures, ExpiryPolicy expiryPolicy)
+            throws ExecutionException, InterruptedException {
+
+        for (int partitionId = 0; partitionId < entriesPerPartition.length; partitionId++) {
+            List<Map.Entry<Data, Data>> entries = entriesPerPartition[partitionId];
+            if (entries != null) {
+                // TODO If there is a single entry, we could make use of a put operation since that is a bit cheaper
+                Operation operation = operationProvider.createPutAllOperation(entries, expiryPolicy, partitionId);
+                Future f = invoke(operation, partitionId, true);
```

Comment:
@serkan-ozal , where do you calculate `putAll` latency? or do you calculate such metric?